### PR TITLE
Fix error messages for :type/:class used in expressions

### PR DESCRIPTION
**Pull request by errael**

### CODE REVIEW

File: src/eval.c [hunk 1]
```diff
@@ -1890,10 +1890,16 @@ set_var_lval(
...
 	    {
+		//if (di != NULL && di->di_tv.v_type == VAR_CLASS)
+		//    abort();
 		if (di != NULL && di->di_tv.v_type == VAR_TYPEALIAS)
 		{
-		    semsg(_(e_cannot_modify_typealias),
-					di->di_tv.vval.v_typealias->ta_name);
+		    // TODO: use standard message, note that if a class type
+		    //	     is caught here, it does not have a class object
+		    //	     so no name.
```
Comment (yegappan) [inline src/eval.c]: The class name is available in `di->di_tv.vval.v_class->class_name`.

File: src/evalvars.c [hunk 2]
```diff
@@ -3974,8 +3974,10 @@ set_var_const(
 		goto failed;
 	    }
 
+	    //if (check_typval_is_value(&di->di_tv) == FAIL) // TODO: class is null
 	    if (di->di_tv.v_type == VAR_TYPEALIAS)
 	    {
+		check_typval_is_value(&di->di_tv);
 		semsg(_(e_cannot_modify_typealias),
```
Comment (yegappan) [inline src/evalvars.c]: This semsg() can be removed?

File: src/evalvars.c [hunk 3]
```diff
@@ -3974,8 +3974,10 @@ set_var_const(
 		goto failed;
 	    }
 
+	    //if (check_typval_is_value(&di->di_tv) == FAIL) // TODO: class is null
```
Comment (yegappan) [inline src/evalvars.c]: Under what conditions is di->di_tv.vval.v_class NULL?

File: src/vim9type.c [hunk 4]
```diff
@@ -1871,7 +1871,11 @@ check_typval_is_value(typval_T *tv)
 	return OK;
     if (tv->v_type == VAR_CLASS)
     {
-        semsg(_(e_using_class_as_value_str), tv->vval.v_class->class_name);
+	if (tv->vval.v_class != NULL)
+	    semsg(_(e_using_class_as_value_str), tv->vval.v_class->class_name);
+	else
+	    //abort(); // How is there a typval without a v_class?
```
Comment (yegappan) [inline src/vim9type.c]: If null_class is used, then v_class will be null.

**Comment by errael** [comment]
Turns out after preventing high level use of :type/:class as values. Many expression situations allowed them. This PR takes care of many of the leftovers. There's probably more out there.

**Comment by errael** [comment]
@chrisbra @yegappan If it is relevant, when considering if this PR should make it into vim9.1, this PR is all about error messages. It was delayed a few days by discovering/fixing an old bug that interfered.

**Review by yegappan — APPROVED** [review]

**Comment by vim-ml** [comment]
Hi Ernie,

On Wed, Dec 20, 2023 at 10:40 AM errael ***@***.***> wrote:

> @chrisbra <https://github.com/chrisbra> @yegappan
> <https://github.com/yegappan> If it is relevant, when considering if this
> PR should make it into vim9.1, this PR is all about error messages. It was
> delayed a few days by discovering/fixing an old bug that interfered.
>
>
>
The changes look good to me.

Regards,
Yegappan

**Comment by chrisbra** [comment]
thanks, applied

**Pull request closed without merging**